### DISCUSSION TITLE
Macros for ble.sh

### DISCUSSION THREAD

**Discussion by mozirilla213** [akinomyoga/ble.sh#294] [Ideas]
So zsh supports aliases with positional parameters like
```zsh
$ alias mid='echo $1 mid $2'
$ mid one two
one mid two
```
making them a simple form of a macro, but since ble.sh in vi-mode can use the operator filter `!` like in vim, we could process the text in our command line with whatever command and have it expand to a more complex code.

So I'll use the `m4` macro processor for example, it should be available for Unix. Create a file `bash.m4` with the following content and put in let's say in your home directory:

```m4
define(`mid', `echo $1 `mid' $2')dnl
```
Now open bash with ble.sh in vi-mode, write the following, then enter normal mode, type double bang `!!` and enter the m4 command, then press enter:
```bash
$ mid(one, two)
!m4 ~/bash.m4 -
$ echo one mid two
one mid two
```
And that's how to get macro expansions in ble.sh, you could make far more complex examples if you learn some `m4`, or some other macro processor, or make your own tools with whatever language. (Note that writing bad macros in m4 leads to infinite loops)

Now I wish I knew a way to better integrate this so that I could press a keybinding and execute this macro expansion (or whatever command in operator filter mode), like how I can do with abbreviations. It could be yet another killer feature of ble.sh

**Comment by mozirilla213** [comment]
```bash
function ble/widget/my/ex1 {
 ble/widget/vi_nmap/linewise-visual-mode
 sleep .4
 ble/keymap:vi/operator:filter
 ble/widget/insert-string "m4 ~/bash.m4 -"
 #ble/widget/accept-single-line-or-newline
}
```
Doesn't quite work

**Comment by akinomyoga** [comment]
> Now I wish I knew a way to better integrate this so that I could press a keybinding and execute this macro expansion (or whatever command in operator filter mode), like how I can do with abbreviations. It could be yet another killer feature of ble.sh

It seems just achievable with the plain Bash feature to me. Is it different from just doing something like the below?

```bash
bind -x '"\C-t": READLINE_LINE=$(m4 ~/bash.m4 - <<< "$READLINE_LINE")'
```

**Edit** (2023-03-09): I have added the missing `-` as pointed out by @mozirilla213.

**Reply by mozirilla213** [reply]
Oh yeah, but don't forget to put the `-`
```bash
bind -x '"\C-t": READLINE_LINE=$(m4 ~/bash.m4 - <<< "$READLINE_LINE")'
```
But you're right, it's achievable for plain bash.

Now I remember I stopped making `READLINE_LINE` binds for ble.sh because the `$PWD` on vim-airline didn't refresh when doing so, I guess that's why it slipped my mind; but in this case, the PWD is irrelevant so yeah nevermind 😄.
I guess this could be a little blog post for bash tips in general then.

**Reply by akinomyoga** [reply]
> because the `$PWD` on vim-airline didn't refresh when doing so,

To comment on this, the prompts keep their values at the end of the previous user commands unless the update is requested. To track the changes of variables to update the prompt segment immediately, you need to call `ble/prompt/unit/add-hash` in the custom prompt sequence as described in [§4.12.2 User-defined prompt sequences](https://github.com/akinomyoga/ble.sh/wiki/Manual-%C2%A74-Editing#4122-user-defined-prompt-sequences).

This is an example implementation for `$PWD`:

```bash
function ble/prompt/backslash:mypwd {
  ble/prompt/unit/add-hash '$PWD'
  ble/prompt/print "$PWD"
}
bleopt vim_airline_section_c='\q{mypwd}'
```

For example, suppose we have the following `bind -x`, 

```bash
$ bind -x '"\C-t": cd - &>/dev/null'
```

The prompt segments containing `\q{mypwd}` should be updated on pressing <kbd>C-t</kbd>.

**Reply by akinomyoga** [reply]
> Oh yeah, but don't forget to put the `-`

Thank you! I have corrected the code example in the original answer above.
